### PR TITLE
issue: 915305 Peer notification fault tolerance support

### PR DESCRIPTION
**Pull request by igor-ivanov**
@olgk please review

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/726/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/732/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/733/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/734/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/735/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/736/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/739/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/741/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/744/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/745/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/750/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/757/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/759/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/760/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/771/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/772/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/779/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
@OphirMunk @olgk  Do you see any issues with this bug fix preventing these changes to be included in master?

**Comment by olgk** [comment]
@igor-ivanov , please rebase on current master

**Comment by igor-ivanov** [comment]
rebased

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1701/ for details (Mellanox internal link).

File: src/vma/util/agent.cpp [hunk 1]
```diff
@@ -77,12 +77,14 @@ agent* g_p_agent = NULL;
 
 agent::agent() :
 		m_state(AGENT_CLOSED), m_sock_fd(-1), m_pid_fd(-1),
-		m_msg_num(512), m_msg_grow(16)
+		m_msg_num(512), m_msg_grow(16),
+		m_inactive_treshold(10), m_alive_treshold(10)
```
Comment (OphirMunk) [inline src/vma/util/agent.cpp]: 1. Please change all hard coded numbers to be under #define
2. Currently inactive threshold and alive thresholds cannot be configured. Please suggest how to be able to modify them (can be a separate PR)

Comment (igor-ivanov) [inline src/vma/util/agent.cpp]: 1. agree
2. it could be done later when we see that is needed.

File: src/vma/util/agent.cpp [hunk 2]
```diff
@@ -227,18 +230,190 @@ agent::~agent()
...
+			list_del_init(&cb->item);
+			free(cb);
+			m_cb_lock.unlock();
+			return ;
+		}
+	}
+	m_cb_lock.unlock();
+}
+
+int agent::put(const void *data, size_t length, intptr_t tag, void **saveptr)
```
Comment (igor-ivanov) [inline src/vma/util/agent.cpp]: it is true. It is not used for sending tcp state but might be used later.

File: src/vma/util/agent.cpp [hunk 3]
```diff
@@ -227,18 +230,190 @@ agent::~agent()
...
+	}
+
+	m_msg_lock.lock();
+
+	if (saveptr) {
+		msg = (agent_msg_t *)(*saveptr);
+	}
+
+	if ((AGENT_ACTIVE == m_state) ||
+			((saveptr) &&
```
Comment (igor-ivanov) [inline src/vma/util/agent.cpp]: `saveptr` is a part of current agent::put() api. It is not used for this specifc case but it might be used for future.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1547/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1549/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1551/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1552/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1561/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
These changes are moved to https://github.com/Mellanox/libvma/pull/706
